File: bin/src/meminfo.zig
const std_c = @import("std-c");
const std = @import("std");

const sysinfo = std_c.sys.utils.zsysinfo;
const print = std_c.print;
const File = std_c.stdio.File;
const allocator = std_c.heap.c_allocator;

const Mode = enum {
    Bytes,
    KiB,
    MiB,
    Verbose,
};

const MemInfo = struct {
    total: usize,
    free: usize,
    used: usize,
};

pub fn main() !void {
    const info_file = try File.open("proc:/meminfo", .{ .read = true });
    defer info_file.close();

    const info_str = try info_file.reader().readUntilEOF();

    const parsed_info = try std.json.parseFromSlice(MemInfo, allocator, info_str, .{ .allocate = .alloc_if_needed, .ignore_unknown_fields = true });
    defer parsed_info.deinit();

    const info = parsed_info.value;

    var mode: Mode = .Verbose;
    var args = std_c.sys.args();

    while (args.next()) |arg| {
        if (std.mem.eql(u8, arg, "-b")) {
            mode = .Bytes;
        } else if (std.mem.eql(u8, arg, "-k")) {
            mode = .KiB;
        } else if (std.mem.eql(u8, arg, "-m")) {
            mode = .MiB;
        }
    }

    switch (mode) {
        .Bytes => {
            print("{}B/{}B\n", .{ info.used, info.total });
        },
        .KiB => {
            print("{}KiB/{}KiB\n", .{ info.used / 1024, info.total / 1024 });
        },
        .MiB => {
            print("{}MiB/{}MiB\n", .{ info.used / 1024 / 1024, info.total / 1024 / 1024 });
        },

        .Verbose => {
            print("memory info:\n", .{});
            print("{}B used of {}B, {}B usable\n", .{ info.used, info.total, info.free });

            print("{}KiBs used of {}KiBs, {}KiBs usable\n", .{ info.used / 1024, info.total / 1024, info.free / 1024 });

            print("{}MiBs used of {}MiBs, {}MiBs usable\n", .{ info.used / 1024 / 1024, info.total / 1024 / 1024, info.free / 1024 / 1024 });
        },
    }
}

comptime {
    _ = std_c;
}

File: binutils/src/bin/meminfo.rs
use extra::{json::MemInfo, tri_io};
use safa_api::errors::SysResult;

enum Mode {
    Bytes,
    KiB,
    MiB,
}

impl Default for Mode {
    fn default() -> Self {
        Self::MiB
    }
}

fn main() -> SysResult {
    let mut verbose = true;
    let mut mode = Mode::default();

    for arg in std::env::args() {
        match arg.as_str() {
            "-r" | "--raw" => verbose = false,
            "-k" => mode = Mode::KiB,
            "-b" => mode = Mode::Bytes,
            "-m" => mode = Mode::MiB,
            _ => {}
        }
    }

    let info = tri_io!(MemInfo::fetch());
    if !verbose {
        match mode {
            Mode::Bytes => println!("{}Bs/{}Bs", info.used(), info.total()),
            Mode::KiB => println!("{}KiBs/{}KiBs", info.used_kib(), info.total_kib()),
            Mode::MiB => println!("{}MiBs/{}MiBs", info.used_mib(), info.total_mib()),
        }
    } else {
        match mode {
            Mode::Bytes => println!(
                "{}Bs used of {}Bs, {}Bs free",
                info.used(),
                info.total(),
                info.free()
            ),
            Mode::KiB => println!(
                "{}KiBs used of {}KiBs, {}KiBs free",
                info.used_kib(),
                info.total_kib(),
                info.free_kib()
            ),
            Mode::MiB => println!(
                "{}MiBs used of {}MiBs, {}MiBs free",
                info.used_mib(),
                info.total_mib(),
                info.free_mib()
            ),
        }
    }

    SysResult::Success
}
